File: 4.Authentication and authorization for backend engineers/code/typescript/08_stateless_authentication_7.ts
import { Request, Response, NextFunction } from 'express';
import jwt from 'jsonwebtoken';

interface User {
    id: string;
    role: string;
}

const secret = 'keep-this-very-secret';

// verify signature + expiry; returns the claims if valid
function verify(tokenStr: string): any {
    try {
        return jwt.verify(tokenStr, secret, { algorithms: ['HS256'] });
    } catch (err) {
        throw new Error('Invalid signature or expired token');
    }
}

export function requireJWT(req: Request, res: Response, next: NextFunction): void {
    const authHeader = req.headers.authorization;
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
        res.status(401).send('authentication failed');
        return;
    }
    
    try {
        const token = authHeader.substring(7);
        const claims = verify(token);
        
        // no store lookup ,  identity comes from the verified token
        const u: User = { id: claims.sub, role: claims.role };
        
        (req as any).user = u;
        next();
    } catch (err) {
        res.status(401).send('authentication failed');
    }
}

File: 4.Authentication and authorization for backend engineers/code/rust/08_stateless_authentication_7.rs
use actix_web::{dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform}, Error, HttpMessage};
use futures_util::future::{ok, LocalBoxFuture, Ready};
use std::rc::Rc;
use jsonwebtoken::{decode, Validation, DecodingKey};
use serde::{Deserialize, Serialize};

const SECRET: &[u8] = b"keep-this-very-secret";

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Claims {
    pub sub: String,
    pub role: String,
}

#[derive(Clone)]
pub struct User {
    pub id: String,
    pub role: String,
}

// verify signature + expiry; returns the claims if valid
fn verify(token_str: &str) -> Result<Claims, jsonwebtoken::errors::Error> {
    let mut validation = Validation::new(jsonwebtoken::Algorithm::HS256);
    validation.validate_exp = true;
    
    let token_data = decode::<Claims>(
        token_str,
        &DecodingKey::from_secret(SECRET),
        &validation,
    )?;
    
    Ok(token_data.claims)
}

pub struct RequireJwt;

impl<S, B> Transform<S, ServiceRequest> for RequireJwt
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Transform = RequireJwtMiddleware<S>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ok(RequireJwtMiddleware { service: Rc::new(service) })
    }
}

pub struct RequireJwtMiddleware<S> {
    service: Rc<S>,
}

impl<S, B> Service<ServiceRequest> for RequireJwtMiddleware<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    forward_ready!(service);

    fn call(&self, req: ServiceRequest) -> Self::Future {
        let auth_header = req.headers().get("Authorization");
        
        if let Some(header) = auth_header {
            if let Ok(header_str) = header.to_str() {
                if header_str.starts_with("Bearer ") {
                    let token = &header_str[7..];
                    if let Ok(claims) = verify(token) {
                        // no store lookup ,  identity comes from the verified token
                        let u = User {
                            id: claims.sub.clone(),
                            role: claims.role.clone(),
                        };
                        req.extensions_mut().insert(u);
                        
                        let srv = self.service.clone();
                        return Box::pin(async move { srv.call(req).await });
                    }
                }
            }
        }
        
        Box::pin(async move {
            Err(actix_web::error::ErrorUnauthorized("authentication failed"))
        })
    }
}
